tomee/tomee-microprofile/mp-common/src/main/java/org/apache/tomee/microprofile/TomEEMicroProfileListener.java

```diff
@@ -122,7 +122,13 @@ public void processApplication(@Observes final BeforeEvent<AfterApplicationCreat
         // index only once and pass it everywhere it's needed. Also in order to build the index, we need the entire
         // application so doing it here from the AppInfo is way simpler
         try {
-            final Index index = of(afterApplicationCreated.getEvent().getApp().libs);
+            final Index index;
+            if(indexCache.containsKey(afterApplicationCreated.getEvent().getApp())) {
+                index = indexCache.get(afterApplicationCreated.getEvent().getApp());
+            } else {
+                index = of(afterApplicationCreated.getEvent().getApp().libs);
+                indexCache.put(afterApplicationCreated.getEvent().getApp(), index);
```

Comment:
you likely want to listen for application undeployment and drop it from the cache cause like that it leaks in the app server

Reply:
Ok, I need some help here, there is no method for this?

Reply:
I could use the AssemblerBeforeApplicationDestroyed event for this but actually the cache is only needed for creation. 
A better approach could be to remove the index from the cache, if the corresponding AppInfo is fully deployed.

Is there a event for this?

Reply:
I'd observe this destroyed event as a guarantee (error case?) but `AssemblerAfterApplicationCreated` for success case should be fine if it is not needed after assembler need

Reply:
AssemblerAfterApplicationCreated seems not to be the event I'm looking for, it is called before AfterApplicationCreated.

Reply:
depends the mode you are in (for webapps it is likely different than for ear).
one trivial mode can be to just add a cdi observer and do it after validation event (+ error case fallback as mentionned earlier)? can be simpler to just follow the lifecycle and use appfinder to lookup the appinfo

side note: i didnt check how relevant it is there (sorry if not but I tend to use another MP flavor) but we tend to attach runtime data (jandex there) to AppContext/webcontext objects this way no need to care of the destruction so it can be saner to add a flag in appinfo to flag it as requiring jandex and compute jandex lazily (in cdi builder?) from info tree data.
It would be stored in AppContext#data (or WebContext#data if relevant).
This way the lifecycle is easier to manage and it is more aligned on the internals but it is a bit more complex to wire.
Since CdiBuilder has both it can be a good place to create jandex if flagged as such from the observer you patched and to clean it up after it was built (so cdi is started, no more need of jandex I guess).

Reply:
I followed another approach. The AppInfo knows all webApps that need to be deployed.
I will remove the cache entry, if the last webApp for the given AppInfo has been deployed.

As suggested, I also listen for the AssemblerBeforeApplicationDestroyed event and will remove the corresponding cache entry, if it still exists